lib/opam_solve.ml

```diff
@@ -251,6 +254,65 @@ module Make_solver (Context : OPAM_MONOREPO_CONTEXT) :
       rolemap []
     |> List.filter_map ~f:Solver.package_name
 
+  let no_matching_versions diagnostics =
+    let rolemap = Solver.diagnostics_rolemap diagnostics in
+    Pkg_map.fold
+      (fun pkg component acc ->
```

Comment:
I would prefer to keep the anonymous function for the `fold` (since it is quite specific that it does it is better to see that the arguments and their types are meant to align with the signature for `fold`) but I've tried to factor out more verbose internal helpers into named functions and give them somewhat understandable names.